[joyngjr] iP

joyngjr opened the pull request:
# **Bond**
> Marriage? Conventional happiness? I got rid of those aspirations as well as my identity, the day I became a spy – Loid Forger 

## _What is Bond?_
Bond is a to-do list management system which is
- text-based
- easy to learn
- fast to use

## _How to use it?_
1. Download files from [here](https://github.com/joyngjr/ip)
2. Run file src/main/duke/Duke.java
3. Refer to list of available user commands (type `help` to get a list of commands)
4. Add/edit tasks!

## _Features_
Experiment with it 🦙 
- [ ] Add tasks (to-do, event, deadline)
- [x] Mark/unmark tasks
- [ ] Delete tasks
- [ ] Keyword search on tasks

## _Commands Available_
```java
public static String generate() {
        return "Here is a list of commands that you may use:\n"
                + "todo <task> - creates a new task\n"
                + "deadline <task> /<yyyy-mm-dd> - creates a task with an end date\n"
                + "event <task> /<yyyy-mm-dd> /<yyyy-mm-dd> - creates a task with start and end date\n"
                + "mark/unmark <int> - change status of task\n"
                + "delete <int> - deletes a task in the list\n"
                + "find <word> - returns tasks in list which matches word\n"
                + "missions - prints out a todo list of existing tasks\n"
                + "help - prints out this list of commmands\n"
                + "bye - to close program\n";
    }
```

Commit:
Add Gradle support

joyngjr committed:
Edit Duke.java

joyngjr committed:
Add records tracker

joyngjr committed:
Add Task.java

joyngjr committed:
Add Deadline.java, Event.java, Todo.java

joyngjr committed:
Add testing files

joyngjr committed:
Add exception handling

joyngjr committed:
Support deletion

joyngjr committed:
Add Commands.java

joyngjr committed:
Add data/duke.txt

joyngjr committed:
Change String to LocalDate

joyngjr committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add data/duke.txt

joyngjr committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Branch was not added when repo was forked

joyngjr committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Change String to LocalDate

# Conflicts:
#	src/main/java/Duke.java

joyngjr committed:
Add more classes for OOP

joyngjr committed:
Move to package duke

joyngjr committed:
Edit gradle set-up

joyngjr committed:
Add JUnit tests

joyngjr committed:
Add javadoc for classes and methods

joyngjr committed:
Modify files to follow coding standards

joyngjr committed:
Add find functionality

joyngjr committed:
Merge branch 'branch-A-JavaDoc' into branch-A-CodingStandard

* branch-A-JavaDoc:
  Add javadoc for classes and methods

# Conflicts:
#	src/main/java/drivers/Parser.java
#	src/main/java/duke/Duke.java
#	src/main/java/task/Task.java

joyngjr committed:
Merge branch 'branch-Level-9' into branch-A-CodingStandard

* branch-Level-9:
  Add find functionality

# Conflicts:
#	data/duke.txt
#	src/main/java/drivers/Parser.java

joyngjr committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Add find functionality
  Modify files to follow coding standards
  Add javadoc for classes and methods

joyngjr committed:
Resolve bugs & improve UI

lywich commented on src/main/java/drivers/Parser.java:
```diff
@@ -0,0 +1,103 @@
+
+            default:
+                System.out.println("Something has gone wrong!");
+                return false;
```
Nice use of switch statements to keep code readable.

lywich commented on src/main/java/support/Commands.java:
```diff
@@ -0,0 +1,4 @@
+package support;
+public enum Commands {
+    bye, missions, unmark, mark, todo, deadline, event, delete, find, help
```
Have you considered capitalizing the commands? Such that a command like "Help" would be able to parse. I recommend using String.toUpperCase()/toLowerCase() method to capitalize/lowercase the words when parsing the input. 

Definitely a toss-up between rejecting commands like "heLP" or having to reject "Help". I respect the decision to only accept lowercase letters but just my two cents.

lywich commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,74 @@
+    // default constructor
+    public void add(Task task) {
+        records.add(task);
+        System.out.println("B: " + task.toString() + "has been added");
```
I recommend adding a whitespace before "has been added". I notice that this might cause an issue whereby it becomes "workhas been added" unless the user themselves add a whitespace. 
```suggestion
        System.out.println("B: " + task.toString() + " has been added");
```

lywich commented on src/main/java/support/Storage.java:
```diff
@@ -0,0 +1,41 @@
+     */
+    public Storage() {
+        try {
+            this.file = new File("./" + "data/duke.txt");
+            file.createNewFile();
```
Perhaps this might break when you delete the data directory? 

Deleting the data directory (that came with the zip file) results in your code being unable to recover from this deletion. Hence, the tasks are no longer stored across sessions. You might be able to observe this if the .jar file is run in 2 separate sessions. 

Just pointing this out as this seems to be a marking criteria. I recommend using Files.createDirectories() and the relative path methods like System.getProperty() and Paths.get() as a way to get around this issue.

lywich reviewed (commented):
In general, very good work. The code quality is very well done apart from some whitespaces (that is really the fault of Intellij) and one bug. I could not find any coding standard violations/unclear code. Everything is readable and design decisions are consistent. Keep up the good work!

joyngjr committed:
Add minimal GUI

joyngjr committed:
Add introduction greeting

joyngjr committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add introduction greeting

abenx162 commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+
+    /**
```
Should the spacing between methods be more consistent? I noticed that there's 2 lines between setComplete and setIncomplete, but none between contains and toString, for example.

abenx162 commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,76 @@
+            if (i != n - 1) {
+                s = s + (i + 1) + ". " + records.get(i).toString() + "\n";
+            } else {
+                s = s + (i + 1) + ". " + records.get(i).toString();
```
Perhaps using a StringBuilder would be more efficient here, as we would not have to create a new String each time a new Task is appended. Although as long as the number of Tasks is small, the difference should not be significant.

abenx162 commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,20 @@
+    public String toString() {
+        return "[E]" + super.toString() +
+                " (from: " + this.start.getMonth() + " " + this.start.getDayOfMonth() + ", " + this.start.getYear() +
+                " to: " + this.end.getMonth() + " " + this.end.getDayOfMonth() + ", " + this.end.getYear() + ")";
```
Perhaps the line breaks should be before the + operators, instead of after?

abenx162 commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,20 @@
+    private LocalDate end;
+
+    public Event(String instruction, String start, String end) {
+        super(instruction.substring(6));
```
I think this line can be hard to understand, especially as the reader may not know that all Event instructions start with the word 'event' (and similarly for the other Task subclasses). Also, might there be issues if the instruction is just "event", and there is no 6th index? Perhaps using instruction.split(" ", 2) might be better to separate out the words.

abenx162 reviewed (commented):
overall very good! just some minor issues, other than that your code is very polished.

joyngjr committed:
Fix data storage issues.

joyngjr committed:
Edit style issues with checkstyle

joyngjr committed:
Merge tag 'A-CheckStyle'

* tag 'A-CheckStyle':
  Edit style issues with checkstyle

joyngjr committed:
Edit closing mechanism for GUI

Lack closing mechanism as users can continue to key in commands after
terminating keyword.

Continued inputs can cause unintended bugs.

Let's make the GUI close itself after detecting the closing command.

Closing the system automatically is safer than waiting for users to
close the application.

joyngjr committed:
Add assertion checks

Lack assertion checks in code.

Creates potentially dangerous situations where bugs are not detected.

Add assert statements before passing data around program and make sure
impossible situations never occur.

Use assert statements to ensure that situations that are not supposed to
happen do not happen.

joyngjr committed:
Do minor refactoring

Some parts of code is convoluted.

Hard to read.

Split string into variables and add boolean method for checking.

Easier for readers to understand.

joyngjr committed:
Merge pull request #2 from joyngjr/A-CodeQuality

Do minor refactoring

joyngjr committed:
Add file for CI

joyngjr committed:
Edit style issues

joyngjr committed:
More style issues found

joyngjr committed:
Add more refactoring

joyngjr committed:
Add user guide

joyngjr committed:
Fix styling issues.

joyngjr committed:
Edit styling issues in user guide

joyngjr committed:
Update Junit tests
